File: aspnet-in-practice/10/VB/PathCanonicalization/App_Code/PathExtensions.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports System.IO

Public Module PathExtensions
    Public Function CanonicalCombine(ByVal basePath As String, ByVal myPath As String) As String
        If String.IsNullOrEmpty(basePath) OrElse String.IsNullOrEmpty(myPath) Then
            Throw New ArgumentNullException()
        End If

        ' let's decode paramters
        basePath = HttpUtility.UrlDecode(basePath)
        myPath = HttpUtility.UrlDecode(myPath)

        If myPath.IndexOfAny(Path.GetInvalidFileNameChars()) > -1 Then
            Throw New FileNotFoundException("FileName not valid")
        End If

        Dim filePath As String = Path.Combine(basePath, myPath)
        If Not filePath.StartsWith(basePath) Then
            Throw New FileNotFoundException("Path not valid")
        End If

        Return filePath
    End Function
End Module

File: aspnet-in-practice/10/CS/PathCanonicalization/App_Code/PathExtensions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.IO;

public static class PathExtensions
{
	public static string CanonicalCombine(string basePath, string path)
	{
		if (String.IsNullOrEmpty(basePath) || string.IsNullOrEmpty(path))
			throw new ArgumentNullException();

		// let's decode paramters
		basePath = HttpUtility.UrlDecode(basePath);
		path = HttpUtility.UrlDecode(path);

		if (path.IndexOfAny(Path.GetInvalidFileNameChars()) > -1)
			throw new FileNotFoundException("FileName not valid");

		string filePath = Path.Combine(basePath, path);
		if (!filePath.StartsWith(basePath))
			throw new FileNotFoundException("Path not valid");

		return filePath;
	}
}

PathExtensions.CanonicalCombine(basePath, PathValue.Text)